sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -1237,8 +1639,10 @@ public Object restriction() {
               .setTransformId(pTransformId)
               .setInputId(mainInputId)
               .setElement(bytesOut.toByteString());
-      // We don't want to change the output watermarks or set the checkpoint resume time since
-      // that applies to the current window.
+      if (!outputWatermarkMapForUnprocessedWindows.isEmpty()) {
```

Comment:
putAllOutputWatermarks should do nothing if the input map is empty so the `if` check is extraneous.